src/odemis/acq/stitching/_registrar.py

```diff
@@ -320,180 +320,135 @@ def _estimateMatch(self, imageA, imageB, shift=(0, 0)):
-                prev_tile = self.tiles[row][col - 1]
-                x, y = self._get_shift(prev_tile, tile, exp_shift)
-            elif xdir == RIGHT:
-                prev_pos = self.shifts[row][col + 1]
-                prev_tile = self.tiles[row][col + 1]
-                x, y = self._get_shift(tile, prev_tile, exp_shift)
-                y = -y
-                x = -x
-            match = self._estimateMatch(
-                prev_tile, tile, (exp_shift[0] - x, exp_shift[1] - y))
-
-            if match < BEST_MATCH and (row > 1 or col > 1):
-                # if the shift calculated does not give a sufficient match,
-                # try the horizontal mean shift along this row and column if
-                # possible
-                try:
-                    mean_x, mean_y = self._mean_shift(row, col, self.coords_hor)
-                except ValueError:
-                    # if a mean value is not available try the last horizontal
-                    # shift measured
-                    mean_x, mean_y = self.last_hor
-                match_c = self._estimateMatch(prev_tile, tile,
-                                              (exp_shift[0] - mean_x, exp_shift[1] - mean_y))
-                if match_c > match:
+        if xdir == LEFT_TO_RIGHT:
+            exp_shift = (int(self.size[1] - self.osize), 0)
+            prev_pos = self.registered_positions[row][col - 1]
+            prev_tile = self.tiles[row][col - 1]
+        elif xdir == RIGHT_TO_LEFT:
+            exp_shift = (-int(self.size[1] - self.osize), 0)
+            prev_pos = self.registered_positions[row][col + 1]
+            prev_tile = self.tiles[row][col + 1]
+        else:
+            raise ValueError("xdir argument is %s, must be either LEFT_TO_RIGHT or RIGHT_TO_LEFT." % xdir)
+        x, y = self._get_shift(prev_tile, tile, exp_shift)
+
+        # If the quality of the cross-correlation is low, use fallback shift
+        match = self._estimateMatch(prev_tile, tile, (exp_shift[0] - x, exp_shift[1] - y))
+        if match < MIN_MATCH:
+            logging.debug('Horizontal registration results in bad match. Using fallback position.')
+            # Use mean shift of all previously registered tiles in horizontal direction. If none are
+            # available, fall back to expected position (i.e. zero shift)
+            known_shifts_x = filter(None, self.shifts_hor[row])
+            if known_shifts_x:
+                mean_x = numpy.mean(filter(None, self.shifts_hor[row]), axis=0)[0]
+            else:
+                mean_x = 0
+            if filter(None, [s[1] for s in self.shifts_hor]):
+                mean_y = numpy.mean(filter(None, [s[1] for s in self.shifts_hor]), axis=0)[1]
+            else:
+                mean_y = 0
+            # Keep new shift if quality of registration is better
+            match_mean = self._estimateMatch(prev_tile, tile, (exp_shift[0] - mean_x, exp_shift[1] - mean_y))
+            if match_mean > match:
```

Comment:
Is that every going to execute? The cross-correlation will likely return the shift with the best match, so the match of the mean shift might always be lower